File: internal/integrations/snowflake/sql.go
package snowflake

import (
	"errors"
	"fmt"
	"regexp"
	"strings"
)

// Identifiers: Snowflake resolves unquoted identifiers to upper case and
// keeps quoted ones as written. hallpass stores every name in resolved
// form (the exact characters Snowflake compares) and renders it quoted, so
// that a name built from caller input can never change the shape of a
// statement.

var unquotedRe = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_$]{0,254}$`)

// parseIdentifier parses one identifier part, quoted or not, into its
// resolved form.
func parseIdentifier(s string) (string, error) {
	if unquotedRe.MatchString(s) {
		return strings.ToUpper(s), nil
	}
	if len(s) >= 3 && s[0] == '"' && s[len(s)-1] == '"' {
		inner := strings.ReplaceAll(s[1:len(s)-1], `""`, "\x00")
		if strings.Contains(inner, `"`) {
			return "", errors.New("a quoted identifier has an unescaped quote")
		}
		inner = strings.ReplaceAll(inner, "\x00", `"`)
		if inner == "" || len(inner) > 255 {
			return "", errors.New("a quoted identifier is empty or longer than 255 characters")
		}
		for _, r := range inner {
			if r < 0x20 || r == 0x7f {
				return "", errors.New("a quoted identifier carries a control character")
			}
		}
		return inner, nil
	}
	return "", fmt.Errorf("%q is not an identifier (letters, digits, _ and $, or double-quoted)", s)
}

// splitName splits a dotted name into its parts, honouring quotes.
func splitName(s string) ([]string, error) {
	var parts []string
	var cur strings.Builder
	inQuote := false
	for i := 0; i < len(s); i++ {
		c := s[i]
		switch {
		case c == '"':
			inQuote = !inQuote
			cur.WriteByte(c)
		case c == '.' && !inQuote:
			parts = append(parts, cur.String())
			cur.Reset()
		default:
			cur.WriteByte(c)
		}
	}
	if inQuote {
		return nil, errors.New("unterminated quoted identifier")
	}
	parts = append(parts, cur.String())
	return parts, nil
}

// parseName parses a name of exactly n dotted parts into resolved parts.
func parseName(s string, n int) ([]string, error) {
	raw, err := splitName(strings.TrimSpace(s))
	if err != nil {
		return nil, err
	}
	if len(raw) != n {
		return nil, fmt.Errorf("expected %d dotted part(s), found %d", n, len(raw))
	}
	out := make([]string, 0, n)
	for _, p := range raw {
		id, err := parseIdentifier(p)
		if err != nil {
			return nil, err
		}
		out = append(out, id)
	}
	return out, nil
}

// quote renders a resolved identifier for a statement.
func quote(id string) string {
	return `"` + strings.ReplaceAll(id, `"`, `""`) + `"`
}

// quoteName renders resolved parts as a dotted quoted name.
func quoteName(parts []string) string {
	q := make([]string, len(parts))
	for i, p := range parts {
		q[i] = quote(p)
	}
	return strings.Join(q, ".")
}

// sameName reports whether two resolved names are equal.
func sameName(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// stringLiteral renders s as a single-quoted SQL string literal.
func stringLiteral(s string) string {
	s = strings.ReplaceAll(s, `\`, `\\`)
	s = strings.ReplaceAll(s, `'`, `''`)
	return "'" + s + "'"
}

// likeLiteral renders s for SHOW ... LIKE: the wildcards % and _ are
// escaped so the pattern is the literal name.
func likeLiteral(s string) string {
	s = strings.ReplaceAll(s, `\`, `\\`)
	s = strings.ReplaceAll(s, `%`, `\%`)
	s = strings.ReplaceAll(s, `_`, `\_`)
	return stringLiteral(s)
}

File: hallpass-py/src/hallpass/integrations/snowflake/sql.py
"""Identifiers and literals for SHOW statements.

Snowflake resolves unquoted identifiers to upper case and keeps quoted ones
as written. hallpass stores every name in resolved form (the exact
characters Snowflake compares) and renders it quoted, so that a name built
from caller input can never change the shape of a statement.
"""

from __future__ import annotations

import re
from collections.abc import Sequence

from hallpass.core.errors import go_quote, go_trim_space

__all__ = ["like_literal", "parse_identifier", "parse_name", "quote", "quote_name", "same_name", "split_name", "string_literal"]

# (Used with fullmatch: Go's $ does not match before a trailing newline.)
_UNQUOTED_RE = re.compile(r"[A-Za-z_][A-Za-z0-9_$]{0,254}")


def _nbytes(s: str) -> int:
    return len(s.encode("utf-8", "surrogatepass"))


def parse_identifier(s: str) -> str:
    """One identifier part, quoted or not, in its resolved form; raise
    ValueError."""
    if _UNQUOTED_RE.fullmatch(s):
        return s.upper()  # ASCII only, by the pattern
    if _nbytes(s) >= 3 and s.startswith('"') and s.endswith('"'):
        inner = s[1:-1].replace('""', "\x00")
        if '"' in inner:
            raise ValueError("a quoted identifier has an unescaped quote")
        inner = inner.replace("\x00", '"')
        if inner == "" or _nbytes(inner) > 255:
            raise ValueError("a quoted identifier is empty or longer than 255 characters")
        for c in inner:
            if c < "\x20" or c == "\x7f":
                raise ValueError("a quoted identifier carries a control character")
        return inner
    raise ValueError(f"{go_quote(s)} is not an identifier (letters, digits, _ and $, or double-quoted)")


def split_name(s: str) -> list[str]:
    """Split a dotted name into its parts, honouring quotes."""
    parts: list[str] = []
    cur: list[str] = []
    in_quote = False
    for c in s:
        if c == '"':
            in_quote = not in_quote
            cur.append(c)
        elif c == "." and not in_quote:
            parts.append("".join(cur))
            cur = []
        else:
            cur.append(c)
    if in_quote:
        raise ValueError("unterminated quoted identifier")
    parts.append("".join(cur))
    return parts


def parse_name(s: str, n: int) -> tuple[str, ...]:
    """A name of exactly n dotted parts, in resolved parts."""
    raw = split_name(go_trim_space(s))
    if len(raw) != n:
        raise ValueError(f"expected {n} dotted part(s), found {len(raw)}")
    return tuple(parse_identifier(p) for p in raw)


def quote(id: str) -> str:
    """A resolved identifier rendered for a statement."""
    return '"' + id.replace('"', '""') + '"'


def quote_name(parts: Sequence[str]) -> str:
    """Resolved parts as a dotted quoted name."""
    return ".".join(quote(p) for p in parts)


def same_name(a: Sequence[str] | None, b: Sequence[str] | None) -> bool:
    """Whether two resolved names are equal."""
    return tuple(a or ()) == tuple(b or ())


def string_literal(s: str) -> str:
    """s as a single-quoted SQL string literal."""
    s = s.replace("\\", "\\\\")
    s = s.replace("'", "''")
    return "'" + s + "'"


def like_literal(s: str) -> str:
    """s for SHOW ... LIKE: the wildcards % and _ are escaped so the pattern
    is the literal name."""
    s = s.replace("\\", "\\\\")
    s = s.replace("%", "\\%")
    s = s.replace("_", "\\_")
    return string_literal(s)
